### ISSUE TITLE
lost glusterfs endpoints 

### ISSUE THREAD

**Issue by pstinghua** [openshift/origin#6070]

**Comment by pstinghua** [comment]
```
8:26:22 AM	docker-registry-2-1h9kj	Pod	FailedMount	Unable to mount volumes for pod "docker-registry-2-1h9kj_default": unsupported volume type (4063 times in the last 11 hours, 17 minutes)
8:26:22 AM	docker-registry-2-1h9kj	Pod	FailedSync	Error syncing pod, skipping: unsupported volume type (4063 times in the last 11 hours, 17 minutes)
```

**Comment by CuriousA62A2A** [comment]
I do believe that I've seen the same issue.
1. I create a glusterfs endpoint
2. I deploy pods that use the endpoint for persistant storage
3. The day after, if I redeploy the pods they won't start with the above error message.

Recreating the glusterfs endpoint fixes it temporarily.

What data do I need to supply to make this viable for troubleshooting?

**Comment by pmorie** [comment]
@markturansky @rootfs @childsb

**Comment by gravis** [comment]
We're having the same issue here. These endpoints are used in 2 projects, and both have lost the endpoints.

**Comment by gravis** [comment]
We're using 1.0.8 btw.

**Comment by CuriousA62A2A** [comment]
We're on 1.0.8 too

**Comment by vbehar** [comment]
same here, also 1.0.8

**Comment by smarterclayton** [comment]
Upstream https://github.com/kubernetes/kubernetes/issues/13511 may be related - what version of Gluster are you using?  @rootfs

**Comment by smarterclayton** [comment]
We need to also verify the endpoints controller didn't start deleting endpoints that don't have a service.

**Comment by rootfs** [comment]
@gravis @pstinghua @CuriousA62A2A @vbehar what version of glusterfs you are using? Do you see anything in glusterfs logs? The log file name is glusterfs.log in /var/

**Comment by gravis** [comment]
```
||/ Name                                                                                                 Version                                                 Architecture
+++-====================================================================================================-=======================================================-===================
ii  glusterfs-client                                                                                     3.5.2-2+deb8u1                                          amd64
ii  glusterfs-common                                                                                     3.5.2-2+deb8u1                                          amd64
ii  glusterfs-server                                                                                     3.5.2-2+deb8u1                                          amd64
```

On the server, and on the nodes:

```
Installed Packages
glusterfs.x86_64                                                                                                                                                                                                         3.5.6-1.fc21
glusterfs-fuse.x86_64                                                                                                                                                                                                    3.5.6-1.fc21
glusterfs-libs.x86_64                                                                                                                                                                                                    3.5.6-1.fc21
```

I haven't found anything in the logs yet.

**Comment by rootfs** [comment]
as far as I can see there are two probable Kubernetes related issues:
- possible missing glusterfs spec glusterfs mount helper that produced "unsupported volume type". [Here](https://github.com/openshift/origin/blob/v1.0.8/Godeps/_workspace/src/k8s.io/kubernetes/pkg/volume/glusterfs/glusterfs.go#L58) is the detail. @pstinghua can you check if your openshit nodes have glusterfs client package installed and what is your pod like?

- the glusterfs endpoint was deleted. @CuriousA62A2A @gravis when you have the issue, are you able to see the endpoints? `oc get endpoints` 

There is a possible Glusterfs related issue:
- In the upstream kubernetes issue, the FUSE endpoints was disconnected, indicating glusterfs fuse daemon exit unexpectedly. This has to be further diagnosed by looking at the glusterfs logs.

**Comment by gravis** [comment]
the glusterfs endpoint was deleted. @CuriousA62A2A @gravis when you have the issue, are you able to see the endpoints? => No, the endpoints were missing.
Just a guess, we've seen this after a lot of failing deploys, during tests. Since our tests are over, the endpoints haven't moved.

**Comment by rootfs** [comment]
let me dig into this unplanned endpoint deletion issue.

**Comment by smarterclayton** [comment]
Found it - see checkLeftoverEndpoints in endpoints_controller.  Was added
in b49dd0ad1e95

For now, you'll need to create a service with selector = null (not empty
map) to preserve the endpoints.

**Comment by CuriousA62A2A** [comment]
@rootfs I'll chime in with what @gravis said. The endpoints are gone.

**Comment by CuriousA62A2A** [comment]
@smarterclayton good find. I'm not exactly sure that I get it though.
Do I need to create a dummy service with 'selector = null' that consumes a resource from the endpoint to keep it alive or has I got it all backwards?

**Comment by smarterclayton** [comment]
Yes - you'd need to have service "glusterfs" with a null selector in order
for endpoints "glusterfs" to not get deleted automatically.  We definitely
need to update our gluster integration documentation.

**Comment by gravis** [comment]
I have read that somewhere, but I don't find the doc...

**Comment by gravis** [comment]
https://github.com/kubernetes/kubernetes/tree/master/examples/glusterfs

"We need also create a service for this endpoints, so that the endpoints will be persistented."

**Comment by gravis** [comment]
with the obvious issue that people with permissions on services can remove accidentally the endpoints...

**Comment by rootfs** [comment]
@gravis @CuriousA62A2A @pstinghua  do you have a chance to try the [glusterfs service ](https://github.com/kubernetes/kubernetes/blob/master/examples/glusterfs/glusterfs-service.json) to persist the endpoints?

**Comment by gravis** [comment]
I have installed it since. 
I'm not super conformable with leaving this open to users. Glusterfs is clearly underneath infrastructure stuff, I just hope no one will remove this service now :(

**Comment by rootfs** [comment]
@gravis I agree this is not **the** most optimal solution, I am looking for a better fix.

**Comment by gravis** [comment]
Awesome. 
Thanks :)

**Comment by smarterclayton** [comment]
This is similar to the router and registry - usually we recommend managing
this as a permissions problem, where concern about change to the system is
addressed via access control or stronger config management tooling X

On Dec 2, 2015, at 9:19 AM, Philippe Lafoucrière <notifications@github.com>
wrote:

I have installed it since.
I'm not super conformable with leaving this open to users. Glusterfs is
clearly underneath infrastructure stuff, I just hope no one will remove
this service now :(

—
Reply to this email directly or view it on GitHub
<https://github.com/openshift/origin/issues/6070#issuecomment-161310670>.

**Comment by CuriousA62A2A** [comment]
I haven't had time to test yet. Will test early next week at the latest.

**Comment by danmcp** [comment]
@rootfs Can you get the appropriate docs changes for this in place?

**Comment by rootfs** [comment]
@danmcp yes, doc fix is merged via #6167

**Comment by rafalkasa** [comment]
To persist the Gluster endpoints, you also need to create a service.

apiVersion: v1
kind: Service
metadata:
  name: gluster-cluster 
spec:
  ports:
  - port: 1
Save the service definition to a file, for example gluster-service.yaml, then create the endpoints object

https://docs.openshift.com/container-platform/3.7/install_config/storage_examples/gluster_example.html
